File: DemoMethods/src/main/java/net/ravendb/demo/indexes/NameAndCountry.java
package net.ravendb.demo.indexes;

import com.mysema.query.annotations.QueryEntity;
import net.ravendb.abstractions.indexing.FieldIndexing;
import net.ravendb.abstractions.indexing.FieldStorage;
import net.ravendb.client.indexes.AbstractMultiMapIndexCreationTask;

public class NameAndCountry extends AbstractMultiMapIndexCreationTask {

    @QueryEntity
    public static class Result {
        private String name;
        private String country;
        private String id;

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }

        public String getCountry() {
            return country;
        }

        public void setCountry(String country) {
            this.country = country;
        }

        public String getId() {
            return id;
        }

        public void setId(String id) {
            this.id = id;
        }
    }

    public NameAndCountry() {
        QNameAndCountry_Result r = QNameAndCountry_Result.result;

        addMap("from e in docs.Employees " +
                "select new { " +
                "  Name = (e.FirstName + \" \") + e.LastName, " +
                "  Country = e.Address.Country " +
                "}");

        addMap("from c in docs.Companies " +
                "select new { " +
                "  Name = c.Name, " +
                "  Country = c.Address.Country " +
                "}");

        addMap("from s in docs.Suppliers " +
                "select new { " +
                "  Name = s.Name, " +
                "  Country = s.Address.Country " +
                "}");

        index(r.country, FieldIndexing.ANALYZED);
        store(r.name, FieldStorage.YES);
        store(r.name, FieldStorage.YES);
    }
}

File: DemoServer/Indexes/NameAndCountry.cs
using System.Linq;
using DemoServer.Entities;
using Raven.Client.Documents.Indexes;

namespace DemoServer.Indexes
{
    public class NameAndCountry : AbstractMultiMapIndexCreationTask<NameAndCountry.Result>
    {
        public class Result
        {
            public string Name { get; set; }
            public string Country { get; set; }
            public string Id { get; set; }
        }


        public NameAndCountry()
        {
            AddMap<Employee>(employees => from e in employees
                                          select new Result
                                          {
                                              Name = e.FirstName + " " + e.LastName,
                                              Country = e.Address.Country
                                          });



            AddMap<Company>(companies => from c in companies
                                         select new Result
                                         {
                                             Name = c.Name,
                                             Country = c.Address.Country
                                         });

            AddMap<Supplier>(suppliers => from s in suppliers
                                          select new Result
                                          {
                                              Name = s.Name,
                                              Country = s.Address.Country
                                          });


            Index(x => x.Country, FieldIndexing.Search);

            Store(x => x.Name, FieldStorage.Yes);
            Store(x => x.Country, FieldStorage.Yes);
        }
    }

    public class NameAndCountry_JavaScript : AbstractJavaScriptIndexCreationTask
    {
        public class Result
        {
            public string Name { get; set; }
            public string Country { get; set; }
            public string Id { get; set; }
        }

        public NameAndCountry_JavaScript()
        {
            Maps.Add(
                @"map('Employees', (e) => {
                    return {
                        Name : e.FirstName + ' ' + e.LastName,
                        Country : e.Address.Country
                    };
                })"
            );

            Maps.Add(
                @"map('Companies', (c) => {
                    return {
                        Name : c.Name,
                        Country : c.Address.Country
                    };
                })"
            );

            Maps.Add(
                @"map('Suppliers', (s) => {
                    return {
                        Name : s.Name,
                        Country : s.Address.Country
                    };
                })"
            );

            Fields.Add("Country", new IndexFieldOptions
            {
                Indexing = FieldIndexing.Search,
                Storage = FieldStorage.Yes
            });

            Fields.Add("Name", new IndexFieldOptions
            {
                Storage = FieldStorage.Yes
            });
        }
    }
}
